src/storage/src/hummock/compaction_group_client.rs

```diff
@@ -24,27 +24,41 @@ use tokio::sync::RwLock;
 
 use crate::hummock::{HummockError, HummockResult};
 
-/// `CompactionGroupClient` maintains compaction group metadata cache.
-pub struct CompactionGroupClient {
+#[async_trait::async_trait]
+pub trait CompactionGroupClient: Send + Sync + 'static {
+    /// Fast path.
+    async fn try_get_compaction_group_id(&self, prefix: Prefix) -> Option<CompactionGroupId>;
+    /// Slow path.
+    async fn get_compaction_group_id(
+        &self,
+        prefix: Prefix,
+    ) -> HummockResult<Option<CompactionGroupId>>;
+}
+
+/// `CompactionGroupClientImpl` maintains compaction group metadata cache.
+pub struct CompactionGroupClientImpl {
     inner: RwLock<CompactionGroupClientInner>,
     hummock_meta_client: Arc<dyn HummockMetaClient>,
 }
 
-impl CompactionGroupClient {
+impl CompactionGroupClientImpl {
     pub fn new(hummock_meta_client: Arc<dyn HummockMetaClient>) -> Self {
         Self {
             inner: Default::default(),
             hummock_meta_client,
         }
     }
+}
 
+#[async_trait::async_trait]
+impl CompactionGroupClient for CompactionGroupClientImpl {
     /// Tries to get from local cache
-    pub async fn try_get_compaction_group_id(&self, prefix: Prefix) -> Option<CompactionGroupId> {
+    async fn try_get_compaction_group_id(&self, prefix: Prefix) -> Option<CompactionGroupId> {
```

Comment:
Do not `async RwLock` because any RPC request will block other read request.

Reply:
suggest `parking_lot::RwLock` to reduce lock wait.

Reply:
```rust
        let mut guard = self.inner.read();
        if let Some(compaction_group_id) = guard.get(&prefix) {
            return Ok(Some(compaction_group_id));
        }
       drop(guard);
        let compaction_groups = self
            .hummock_meta_client
            .get_compaction_groups()
            .await
            .map_err(HummockError::meta_error)?;
                let mut guard = self.inner.write();
        guard.set_index(compaction_groups);
        Ok(guard.get(&prefix))

```

Reply:
Then we also need to avoid duplicate RPCs.

Reply:
No. This RPC is light and cost little CPU of meta-service. So we do not need to avoid duplicate RPCs